### PR TITLE
feat(otlp): add integration test setup for traces

### LINKED ISSUE #1327: Add integration test for OTLP exporters
In order to increase our confidence when releasing a new version. We should have integration tests for important exporters(OTLP, jager, etc). 

Currently we have limited integration test for jaeger, and CI setup to run integration tests on PRs. But we want to expand/move it to OTLP exporters

### PR DESCRIPTION
**Pull request by TommyCpp**
#1327 
Design discussion issue (if applicable) #

## Changes
- remove jaeger integration test
- add integration test setup.

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by cijothomas** [comment]
Thanks @TommyCpp ! Will you be removing the existing jaeger based integration test in favor of this?

File: opentelemetry-otlp/tests/integration_test/src/asserter.rs
Comment (cijothomas) [inline opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: we can do similar exception for spanid/traceid as well, to keep this simpler.

Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: I think for span ID and trace ID I'd rather use a `IncrementIdGenerator` so that we get predictable IDs. Because we want to make sure the parent-child relationship is correct

File: opentelemetry-otlp/tests/integration_test/expected/traces.json
Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/expected/traces.json:36]: Good point I will add it

**Comment by TommyCpp** [comment]
> Will you be removing the existing jaeger based integration test in favor of this?

I think it's fine to keep the integration test in jaeger in place? That only tests jaeger so it's not mutally exclusive. 

I plan to add a jaeger container here to integrate test otlp exporter w/ jaeger backend too in follow up PRs

File: opentelemetry-otlp/tests/integration_test/src/asserter.rs
Comment (cijothomas) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: In my view, that is partially defeating the purpose of integration tests.
Integration tests should be executing same code as a user would be doing, and custom IdGenerator deviates from that. 
Testing parent-child relation is not the goal of integration tests, right? 

Here's what I am thinking:
1. Execute the same hello world style example code + OTLP Exporter
2. Setup OTel Collector and have it save telemetry to file.
3. Validate that the file has expected telemetry. Things like attributes of diff. types are received, span events are received, metrics are correct etc. - You can of course validate that span's parent-child relation is maintained without pre-knowing the actual traceid/spanid.

As a first step, I'd suggest to start with simply validating that a file is created by collector, and we can add validation of its content as a next PR, to keep things moving easily.

Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: Yeah the current setup complete the 1 and 2. 

> Integration tests should be executing same code as a user would be doing, and custom IdGenerator deviates from that.

I think the `IdGenerator` is a small enough component that self-contained and can be well tested so the chance we missed anything bu not using the `RandomIdGenerator` is minimal

> You can of course validate that span's parent-child relation is maintained without pre-knowing the actual traceid/spanid.

Yeah it's boils down to a compare two forests. But I didn't find a easy way to write the expected span tress before asserting. I think a prediable ID is a good start point for now until we start testing context progatation through multiple containers. 

In other words, if we are only testing export from single container to collector. `IncrementIdGenerator` can satisify the requirment to be unique in this simple system. Randomness is not required by spec(though it's preferred)

Comment (cijothomas) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: > Yeah the current setup complete the 1 and 2.


Not quite as this is executing diff. code than user would be (due to idgenerator). IMHO, Intergration tests should treat the library as blackbox only, and dont change anything at all, including idgenerator.

Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: Not sure I understand. 

> the current setup complete the 1 and 2.

The current setup includes the following steps:
- Execute the same hello world style example code + OTLP Exporter. This is done by spin up a collector container using testcontainer create and mount a local file to collector container.
- Setup OTel Collector and have it save telemetry to file. The result will be written to the mounted file by collector container

I think the difference is on how do we assert the result. The current setup opt in a quick appoaches where we parse the json output from collector and examine the `ResourceSpan` is expected, which need the predictable trace IDs and span IDs. 

I agree on the long term we should take a step further and rebuild the span relationship from `ResourceSpan`s but not sure if it should be a blocker for this PR

Comment (cijothomas) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: > examine the ResourceSpan is expected, which need the predictable trace IDs and span IDs.

Why need predictable IDs? We don't use predictable timestamps, right? I was asking to do same for traceid,spanid as well. Anything that is auto-generated (and hence unpredictable), can be avoided from the assertion. Anything user-provided, and hence predictable can be validated.

To reiterate my opinion:
We should not use custom id generator, and instead only execute test app as a normal user would execute. I understand IdGenerator is a small change only, but that is still partially defeating the purpose of integration tests with otlp!

Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: >  I was asking to do same for traceid,spanid as well

I think the integration tests don't add much values if we cannot validate parent-children relationship between spans thus I am hesitiate to just ignore the trace IDs and span IDs. For example, say we have a bug that caused every span to be a root span. Without assertion on span IDs and traces IDs the integration test won't catch this. 

> Why need predictable IDs? We don't use predictable timestamps, right?

One point is we do allow users to provide their own `IdGenerator` while we don't allow users to bring their own clock

Either case if you feel strong about the predictable IdGenerator I can explore other options to assert the span relationships 🙂

Comment (cijothomas) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:4]: > I think the integration tests don't add much values if we cannot validate parent-children relationship

Agree. But we dont need predictable ids for that. We just need to validate that for the root span, `parentSpanId` will be empty, and the `parentSpanId` of the child will be the `spanId` of the parent. The values themselves can be anything! Its totally okay to add this validation in the future.

Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs]: > We just need to validate that for the root span, parentSpanId will be empty, and the parentSpanId of the child will be the spanId of the parent

This only validate the number of root spans and non root spans. To fully validate the span relationship you need to build two span forests and build a map between each spans in expected result and actual results, then validate they are the same. We will also need to handle links, which make the question even harder that we need build and compare two directed cyclic graph.

Prediable ID avoid all those operation with small drawback that we are not testing the random generation part of the operation, which IMO is a good trade off.

**Comment by TommyCpp** [comment]
Looking at the output of the file exporter. I find the timestamp fields are all string while they should be uint64(the example json from the otlp protocol has them as ints). I will open an issue on file exporter side to see if that's expected. Until then probably won't be able to merge this

**Comment by cijothomas** [comment]
> Looking at the output of the file exporter. I find the timestamp fields are all string while they should be uint64(the example json from the otlp protocol has them as ints). I will open an issue on file exporter side to see if that's expected. Until then probably won't be able to merge this

I think we can unblock ourselves by merging this PR just after checking a file is generated. Validating its contents can be done as a next PR, so we can keep things moving.

File: opentelemetry-otlp/tests/integration_test/src/asserter.rs
Comment (cijothomas) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs]: _but we did seen a lot of issues with context propagation_

That should be tested via InMemoryExporter, right? Why its part of integration tests ?

File: opentelemetry-otlp/tests/integration_test/Cargo.toml
Comment (cijothomas) [inline opentelemetry-otlp/tests/integration_test/Cargo.toml:16]: do we use this anywhere?

File: opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs
Comment (cijothomas) [inline opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs:71]: do we need this port?

File: opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs
Comment (cijothomas) [inline opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs:72]: do we need this? I think only 4317, 4318 should be needed for testing the gRPC and Http based exporting and nothing more.

File: opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs
Comment (cijothomas) [inline opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs:88]: mm.. These generally cause random failures in the CI checks...One way to mitigate this is by checking for the file's presence for upto N seconds, sleeping in between checks. (A good candidate for this future improvement.)

File: opentelemetry-otlp/tests/integration_test/tests/traces.rs
Comment (cijothomas) [inline opentelemetry-otlp/tests/integration_test/tests/traces.rs:29]: we don't validate this so, I suggest to remove this customization.

File: opentelemetry-otlp/tests/integration_test/src/images.rs
Comment (lalitb) [inline opentelemetry-otlp/tests/integration_test/src/images.rs:33]: nit - do we need all these ports exposed?

File: opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs
Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs]: Removed both

File: opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs
Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs:86]: We can add retry with backoff once we have more complicate integration tests

File: opentelemetry-otlp/tests/integration_test/src/asserter.rs
Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:1]: I don't have a strong preference but feel like we can refactor as we adding more signals to it

Comment (lalitb) [reply opentelemetry-otlp/tests/integration_test/src/asserter.rs:1]: Agree.

File: opentelemetry-otlp/tests/integration_test/tests/traces.rs
Comment (lalitb) [inline opentelemetry-otlp/tests/integration_test/tests/traces.rs:89]: Is this test meant to validate the TraceAsserter logic, as it is always testing two identical spans?

File: opentelemetry-otlp/tests/integration_test/tests/traces.rs
Comment (lalitb) [inline opentelemetry-otlp/tests/integration_test/tests/traces.rs:73]: Just to confirm, as of now we are not testing the success scenario, only testing the creation of result file?

Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/tests/traces.rs:73]: Yes

File: opentelemetry-otlp/tests/integration_test/tests/traces.rs
Comment (TommyCpp) [reply opentelemetry-otlp/tests/integration_test/tests/traces.rs:89]: Yes. it's like meta testing for testing tool

File: opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs
Comment (cijothomas) [inline opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs:51]: do we only expect to run this manually? Or we expect this to be run in CI as well?

**Review by lalitb — APPROVED** [review]
LGTM, thanks. The validation of exported data can be in subsequent PR.

File: opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs
Comment (cijothomas) [reply opentelemetry-otlp/tests/integration_test/tests/integration_tests.rs:51]: nvm. found that we run even ignored in intergration tests.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**
